### ISSUE TITLE
BUG: Wrong p-values with Wilcoxon signed-rank test because of wrong rank data 

### ISSUE THREAD

**Issue by kentegrate** [scipy/scipy#17667]
### Describe your issue.

# Problem
I realized that the rank data inside the Wilcoxon signed-rank tests are wrong due to floating-point errors, resulting in wrong p-values.

# When it occurs
When I input the data d1 and d2 to the wilcoxon function as shown below, the rank data![(_morestats.py L.3466)](https://github.com/scipy/scipy/blob/main/scipy/stats/_morestats.py#L3466)
 should be [1, 3, 3, 3]. However, [1, 2, 3, 4] was calculated inside the function.
This will lead to wrong p-values in some cases.

# Possible cause of the problem
This problem seems to happen because of the floating-point errors when the difference between data x and y is calculated in the function![_morestats.py L.3433](https://github.com/scipy/scipy/blob/main/scipy/stats/_morestats.py#L3433). 

# Suggestion
I suggest rounding d according to the precision of the dtype of x and y.
Or, are there any modules in scipy that can deal with such issues?

### Reproducing Code Example

```python
>> d1 = np.array([0.5,0.825,0.375,0.5])
>> d2 = np.array([0.525,0.775,0.325,0.55])
>> T, p = stats.wilcoxon(d1, d2, alternative='greater')
```

### Error message

```shell
>> # Printed abs(d) (L.3433) and rank data (L.3466) in stats._morestats.py with np.set_printoptions(precision=200)
abs d: [0.025000000000000022 0.04999999999999993  0.04999999999999999. 0.050000000000000044]
r: [1. 2. 3. 4.]
>>print(T, p)
4.0 1.0
```

### SciPy/NumPy/Python version information

1.9.3 1.23.3 sys.version_info(major=3, minor=10, micro=2, releaselevel='final', serial=0)

**Comment by mdhaber** [comment]
Yes, `wilcoxon` "exact" p-values are not exact when there are ties, but there is no warning or documentation as there is when there are zeros. I'll add the warning/documentation for now. I'll actually fix it with `permutation_test` as part of a larger enhancement for all hypothesis tests (that is, all hypothesis tests for which a permutation test is appropriate).

**Comment by kentegrate** [comment]
Hello, thank you very much for your detailed response. I was really confused by why I get different values from other statistics softwares, but it is very clear now. 
I appreciate your effort for adding the warning/documentation. Please feel free to close this issue when it is ready.

**Comment by mdhaber** [comment]
A few more thoughts.
TLDR: In the meantime, would you like to submit a PR that just produces a warning something like

> `method="exact "` calculates approximate p-values when there are ties; consider using `permutation_true`

when there are ties?

---

Actually, the fact that p-values are not exact when there are ties is documented in the paragraph after the list in the Notes:
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wilcoxon.html

[Here](https://github.com/scipy/scipy/issues/17530#issuecomment-1339792328) is some information about using `permutation_test` with `wilcoxon`.

In that same issue, [here](https://github.com/scipy/scipy/issues/17530#issuecomment-1341626700), I wrote a bit about why it's usually ok to use the p-values even when there are ties, and how I will fix it with `permutation_test`. Additional discussion about fixing it is [here](https://github.com/scipy/scipy/pull/16996#issuecomment-1344661271). It's very easy; I just need to settle on an interface. For `wilcoxon`, speed would increase dramatically if the statistic calculation were natively vectorized.

There was talk in gh-13690 about detecting ties and switching to a normal approximation. I suppose I'm fine producing a warning, but I don't see a need to switch to the normal approximation. Would you like to submit a PR for that - just producing a warning something like

> `method="exact "` calculates approximate p-values when there are ties; consider using `permutation_true`

when there are ties?

**Comment by mdhaber** [comment]
To resolve this issue, I'd be happy to review a PR with some or all of the following:

1. Add a warning in the documentation after the definition of `y`, maybe like:
```
    .. warning::
        When `y` is provided, `scipy.stats.wilcoxon` calculates the test
        statistic based on the ranks of the absolute values of ``d = x - y``.
        Roundoff error in the subtraction can result in elements of ``d`` being
        assigned different ranks even when they would be tied with exact 
        arithmetic. Rather than passing `x` and `y` separately, consider 
        computing the difference ``x - y``, rounding as needed to ensure that 
        only truly unique elements are numerically distinct, and passing the 
        result as `x`, leaving `y` at the default value (None).
```
(Yes, the single/double backticks are correct. They are not always correct in the rest of the `wilcoxon` documentation.)

2. Optionally, we could check for elements of `d` that are nearly identical but numerically distinct and emitting a warning in that case.  One idea is:

```python3
import numpy as np
def _detect_distinct_but_close(d, eps=1e-14):
    # Checks for numerically distinct but nearly identical elements of `d`.
    abs_d = np.abs(d)
    diff = np.diff(np.sort(abs_d))
    small_diff = (diff != 0) & (diff / abs_d[1:] < eps)
    return np.any(small_diff)
```

I'm not sure if that's perfect, though. If `x` and `y` are available, it seems like we should consider the inaccuracy in the calculation of `d = x - y`. I wrote some code for this that involves the `np.maximum(x, y)` when computing the relative difference, but I'm not sure if that's quite right; I'd need to think about it more.

3. Optionally, we could add the case from a top post as an example, and show how the numbers can be rounded so that `wilcoxon` can detect the ties.
